File: backend/models/schedule.py
from .bay import Bay
from .request import Request


class Schedule:
    def __init__(self, date):
        self.date = date
        self.profit = 0
        self.loss = 0
        self.bays = {
            Bay.COMPACT,
            Bay.MEDIUM,
            Bay.FULL_SIZE,
            Bay.CLASS_1_TRUCK,
            Bay.CLASS_2_TRUCK,
            Bay.FREE_STATION,
        }

    def add_appointment(self, request: Request):
        vehicle = request.vehicle
        for bay in self.bays:
            if bay.value == vehicle.type or bay.value == Bay.FREE_STATION.value:
                if bay.add_appointment(request):
                    self.profit += vehicle.charge
                    return True
        self.loss += vehicle.charge
        return False

    def to_dict(self):
        return {
            "date": self.date,
            "profit": self.profit,
            "loss": self.loss,
            "bays": [bay.to_dict() for bay in self.bays],
        }

    @classmethod
    def from_dict(cls, data):
        schedule = cls(data["date"])
        schedule.profit = data["profit"]
        schedule.loss = data["loss"]
        schedule.bays = [Bay.from_dict(bay) for bay in data["bays"]]
        return schedule

File: frontend/src/models/schedule.ts
import { Vehicle } from "./vehicle";

export interface Appointment {
    vehicle: Vehicle;
    start_time: string;
    end_time: string;
}

export interface Bay {
    bay_type: string;
    appointments: Appointment[];
}

export enum BayType {
    COMPACT = "Compact",
    MEDIUM = "Medium",
    FULL_SIZE = "Full-size",
    CLASS_1_TRUCK = "Class 1 truck",
    CLASS_2_TRUCK = "Class 2 truck",
    FREE_STATION = "Any",
}

export function getBayTypeByString(vehicleType: string): BayType {
    for (const value of Object.values(BayType)) {
        if (value.toLowerCase() === vehicleType.toLowerCase()) {
            return value;
        }
    }
    return BayType.FREE_STATION;
}

export interface Schedule {
    date: string;
    profit: number;
    loss: number;
    bays: Bay[];
}